node/core/av-store/src/lib.rs

```diff
@@ -0,0 +1,555 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implements a `AvailabilityStoreSubsystem`.
+//!
+
+#![recursion_limit="256"]
+#![warn(missing_docs)]
+
+use std::collections::HashMap;
+use std::io;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use codec::{Encode, Decode};
+use futures::{select, channel::oneshot, FutureExt};
+use kvdb_rocksdb::{Database, DatabaseConfig};
+use kvdb::{KeyValueDB, DBTransaction};
+
+use polkadot_primitives::v1::{
+	Hash, AvailableData, ErasureChunk, ValidatorIndex,
+};
+use polkadot_subsystem::{
+	FromOverseer, SubsystemError, Subsystem, SubsystemContext, SpawnedSubsystem,
+};
+use polkadot_subsystem::messages::AvailabilityStoreMessage;
+
+const LOG_TARGET: &str = "availability";
+
+mod columns {
+	pub const DATA: u32 = 0;
+	pub const NUM_COLUMNS: u32 = 2;
+}
+
+#[derive(Debug, derive_more::From)]
+enum Error {
+	#[from]
+	Erasure(erasure::Error),
+	#[from]
+	Io(io::Error),
+	#[from]
+	Oneshot(oneshot::Canceled),
+	#[from]
+	Subsystem(SubsystemError),
+}
+
+/// An implementation of the Availability Store subsystem.
+pub struct AvailabilityStoreSubsystem {
+	inner: Arc<dyn KeyValueDB>,
+}
+
+fn available_data_key(candidate_hash: &Hash) -> Vec<u8> {
+	(candidate_hash, 0i8).encode()
+}
+
+fn erasure_chunk_key(candidate_hash: &Hash, index: u32) -> Vec<u8> {
+	(candidate_hash, index, 0i8).encode()
+}
+
+#[derive(Encode, Decode)]
+struct StoredAvailableData {
+	data: AvailableData,
+	n_validators: u32,
+}
+
+/// Configuration for the availability store.
+pub struct Config {
+	/// Cache size in bytes. If `None` default is used.
+	pub cache_size: Option<usize>,
+	/// Path to the database.
+	pub path: PathBuf,
+}
+
+impl AvailabilityStoreSubsystem {
+	/// Create a new `AvailabilityStoreSubsystem` with a given config on disk.
+	pub fn new_on_disk(config: Config) -> io::Result<Self> {
+		let mut db_config = DatabaseConfig::with_columns(columns::NUM_COLUMNS);
+
+		if let Some(cache_size) = config.cache_size {
+			let mut memory_budget = HashMap::new();
+
+			for i in 0..columns::NUM_COLUMNS {
+				memory_budget.insert(i, cache_size / columns::NUM_COLUMNS as usize);
+			}
+			db_config.memory_budget = memory_budget;
+		}
+
+		let path = config.path.to_str().ok_or_else(|| io::Error::new(
+			io::ErrorKind::Other,
+			format!("Bad database path: {:?}", config.path),
+		))?;
+
+		let db = Database::open(&db_config, &path)?;
+
+		Ok(Self {
+			inner: Arc::new(db),
+		})
+	}
+
+	#[cfg(test)]
+	fn new_in_memory(inner: Arc<dyn KeyValueDB>) -> Self {
```

Comment:
why doesn't it create `in_memory` db inside the function, I don't see the store used outside?

Reply:
This is used in testing so the main idea is that it could be useful to have a handle to the db outside of the store so the test could look into it directly; though it is not used at the moment